### PR TITLE
Create N replicas of each other-avatar on the client side

### PR DESCRIPTION
**Pull request by luiscuenca**
This PR allows to multiply by N the number of other avatars present on the domain. The replicas are created only on the client for testing purposes.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12373/HighFidelity-Beta-PR13877-1fd0d7d.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12373/HighFidelity-Beta-PR13877-1fd0d7d.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12373/HighFidelity-Beta-Interface-PR13877-1fd0d7d.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12381/HighFidelity-Beta-PR13877-5654acf.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12381/HighFidelity-Beta-PR13877-5654acf.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12381/HighFidelity-Beta-Interface-PR13877-5654acf.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12381/HighFidelity-Beta-PR13877-5654acf.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12381/HighFidelity-Beta-Interface-PR13877-5654acf.exe

**Comment by luiscuenca** [comment]
## ReplicatorApp
https://hifi-content.s3.amazonaws.com/luis/test_scripts/replicator_app/replicatorApp.js
## TEST PLAN
- Start interface and go to a domain different that the TEST_DOMAIN
- Start <b>ReplicatorApp</b>
- Add the TEST DOMAIN address and select the number N of replicas.
- START TEST should take you to the TEST_DOMAIN and replicate other avatars by N.

## Known issues.
Avatar entities are not replicated.
Changing domains after starting the test can remove MyAvatar.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12394/HighFidelity-Beta-PR13877-ddbadf5.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12394/HighFidelity-Beta-PR13877-ddbadf5.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12394/HighFidelity-Beta-Interface-PR13877-ddbadf5.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12394/HighFidelity-Beta-PR13877-ddbadf5.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12394/HighFidelity-Beta-Interface-PR13877-ddbadf5.exe

File: libraries/avatars/src/AvatarData.cpp [hunk 1]
```diff
@@ -918,7 +918,26 @@ int AvatarData::parseDataFromBuffer(const QByteArray& buffer) {
...
+        const int AVATARS_PER_ROW = 3;
+
+        glm::vec3 offset;
+
+        if (false) {
+            qsrand(static_cast<quint64>(getID().toByteArray().toInt()));
+            float xrand = float((qrand() % ((RANDOM_RADIUS + 1) - 10) + 10) / 10.0f);
+            float yrand = float((qrand() % ((RANDOM_RADIUS + 1) - 10) + 10) / 10.0f);
+            offset = glm::vec3(xrand, 0.0f, yrand);
+        }
```
Comment (ZappoMan) [inline libraries/avatars/src/AvatarData.cpp]: coding standard says `} else {`

File: libraries/avatars/src/AvatarData.cpp [hunk 2]
```diff
@@ -918,7 +918,26 @@ int AvatarData::parseDataFromBuffer(const QByteArray& buffer) {
...
         auto data = reinterpret_cast<const AvatarDataPacket::AvatarGlobalPosition*>(sourceBuffer);
-        auto newValue = glm::vec3(data->globalPosition[0], data->globalPosition[1], data->globalPosition[2]);
+
+        const float SPACE_BETWEEN_AVATARS = 2.0f;
+        const int RANDOM_RADIUS = 100;
+        const int AVATARS_PER_ROW = 3;
+
+        glm::vec3 offset;
+
+        if (false) {
```
Comment (ZappoMan) [inline libraries/avatars/src/AvatarData.cpp]: what's this dead code about?

File: libraries/avatars/src/AvatarHashMap.cpp [hunk 3]
```diff
@@ -223,12 +321,16 @@ void AvatarHashMap::processBulkAvatarTraits(QSharedPointer<ReceivedMessage> mess
...
+                    avatar->processTrait(traitType, traitData);
+                    _replicas.processTrait(avatarID, traitType, traitData);
                     lastProcessedVersions[traitType] = packetTraitVersion;
-                } else {
+                }
+                else {
                     skipBinaryTrait = true;
                 }
-            } else {
+            }
```
Comment (ZappoMan) [inline libraries/avatars/src/AvatarHashMap.cpp]: coding standard says `} else {`

File: libraries/avatars/src/AvatarHashMap.cpp [hunk 4]
```diff
@@ -223,12 +321,16 @@ void AvatarHashMap::processBulkAvatarTraits(QSharedPointer<ReceivedMessage> mess
 
                 // check if this trait version is newer than what we already have for this avatar
                 if (packetTraitVersion > lastProcessedVersions[traitType]) {
-                    avatar->processTrait(traitType, message->read(traitBinarySize));
+                    auto traitData = message->read(traitBinarySize);
+                    avatar->processTrait(traitType, traitData);
+                    _replicas.processTrait(avatarID, traitType, traitData);
                     lastProcessedVersions[traitType] = packetTraitVersion;
-                } else {
+                }
```
Comment (ZappoMan) [inline libraries/avatars/src/AvatarHashMap.cpp]: coding standard says `} else {`

File: libraries/avatars/src/AvatarHashMap.cpp [hunk 5]
```diff
@@ -238,11 +340,16 @@ void AvatarHashMap::processBulkAvatarTraits(QSharedPointer<ReceivedMessage> mess
                 if (packetTraitVersion > processedInstanceVersion) {
                     if (traitBinarySize == AvatarTraits::DELETED_TRAIT_SIZE) {
                         avatar->processDeletedTraitInstance(traitType, traitInstanceID);
-                    } else {
-                        avatar->processTraitInstance(traitType, traitInstanceID, message->read(traitBinarySize));
+                        _replicas.processDeletedTraitInstance(avatarID, traitType, traitInstanceID);
+                    }
```
Comment (ZappoMan) [inline libraries/avatars/src/AvatarHashMap.cpp]: coding standard says `} else {`

File: libraries/avatars/src/AvatarHashMap.cpp [hunk 6]
```diff
@@ -238,11 +340,16 @@ void AvatarHashMap::processBulkAvatarTraits(QSharedPointer<ReceivedMessage> mess
...
+                    else {
+                        auto traitData = message->read(traitBinarySize);
+                        avatar->processTraitInstance(traitType, traitInstanceID, traitData);
+                        _replicas.processTraitInstance(avatarID, traitType, traitInstanceID, traitData);
                     }
                     processedInstanceVersion = packetTraitVersion;
-                } else {
+                }
```
Comment (ZappoMan) [inline libraries/avatars/src/AvatarHashMap.cpp]: coding standard says `} else {`

File: libraries/avatars/src/AvatarHashMap.h [hunk 7]
```diff
@@ -41,6 +41,27 @@
...
+    void processTrait(const QUuid& parentID, AvatarTraits::TraitType traitType, QByteArray traitBinaryData);
+    void processDeletedTraitInstance(const QUuid& parentID, AvatarTraits::TraitType traitType, AvatarTraits::TraitInstanceID instanceID);
+    void processTraitInstance(const QUuid& parentID, AvatarTraits::TraitType traitType,
+                                AvatarTraits::TraitInstanceID instanceID, QByteArray traitBinaryData);
+    void setReplicaCount(int count) { _replicaCount = count; }
+    int getReplicaCount() { return _replicaCount; }
+
+private:
+    std::map<QUuid, std::vector<AvatarSharedPointer>> _replicasMap;
+    int _replicaCount;
```
Comment (ZappoMan) [inline libraries/avatars/src/AvatarHashMap.h]: coding standard says initialization should happen here.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12413/HighFidelity-Beta-PR13877-8ede6f1.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12413/HighFidelity-Beta-PR13877-8ede6f1.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12413/HighFidelity-Beta-Interface-PR13877-8ede6f1.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12413/HighFidelity-Beta-PR13877-8ede6f1.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12413/HighFidelity-Beta-Interface-PR13877-8ede6f1.exe

**Comment by qaAdamSmith82** [comment]
Currently testing

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12433/HighFidelity-Beta-PR13877-c19bd89.apk

**Comment by qaAdamSmith82** [comment]
PASS @conklin94122 

The app works great. Replicates bots and people controlled avatars alike.

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12433/HighFidelity-Beta-PR13877-c19bd89.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12433/HighFidelity-Beta-Interface-PR13877-c19bd89.exe

**Comment by qaAdamSmith82** [comment]
Already merged.

**Pull request merged**